MoreLinq/Interleave.cs

```diff
@@ -45,125 +44,63 @@ public static partial class MoreEnumerable
-                    iterators = null;
-                    var shouldContinue = true;
-                    var consumedIterators = 0;
-                    var iterCount = iteratorList.Count;
+                    var enumerator = sequence.GetEnumerator();
 
-                    while (shouldContinue)
+                    // Immediately dispose enumerators of empty sequences.
+                    if (enumerator.MoveNext())
                     {
-                        // advance every iterator and verify a value exists to be yielded
-                        for (var index = 0; index < iterCount; index++)
-                        {
-                            if (!iteratorList[index].MoveNext())
-                            {
-                                // check if all iterators have been consumed and we can terminate
-                                // or if the imbalance strategy informs us that we MUST terminate
-                                if (++consumedIterators == iterCount || imbalanceStrategy == ImbalancedInterleaveStrategy.Stop)
-                                {
-                                    shouldContinue = false;
-                                    break;
-                                }
-
-                                iteratorList[index].Dispose(); // dispose the iterator sice we no longer need it
-
-                                // otherwise, apply the imbalance strategy
-                                switch (imbalanceStrategy)
-                                {
-                                    case ImbalancedInterleaveStrategy.Pad:
-                                        var newIter = iteratorList[index] = Generate(default(T), x => default).GetEnumerator();
-                                        newIter.MoveNext();
-                                        break;
-
-                                    case ImbalancedInterleaveStrategy.Skip:
-                                        iteratorList.RemoveAt(index); // no longer visit this particular iterator
-                                        --iterCount; // reduce the expected number of iterators to visit
-                                        --index; // decrement iterator index to compensate for index shifting
-                                        --consumedIterators; // decrement consumer iterator count to stay in balance
-                                        break;
-                                }
-
-                            }
-                        }
-
-                        if (shouldContinue) // only if all iterators could be advanced
-                        {
-                            // yield the values of each iterator's current position
-                            foreach (var iterator in iteratorList)
-                                yield return iterator.Current;
-                        }
+                        yield return enumerator.Current;
+                        enumerators.AddLast(enumerator);
```

Comment:
I believe you will leak the resource here if `yield` never returns.

Reply:
Indeed. I will try to add a test for that.